README.md
# fourtwentyjs-devp2p

[![NPM Package](https://img.shields.io/npm/v/fourtwentyjs-devp2p.svg)](https://www.npmjs.org/package/fourtwentyjs-devp2p)
[![Actions Status](https://github.com/420integrated/fourtwentyjs-devp2p/workflows/Build/badge.svg)](https://github.com/420integrated/fourtwentyjs-devp2p/actions)
[![Coverage Status](https://img.shields.io/coveralls/420integrated/fourtwentyjs-devp2p.svg)](https://coveralls.io/r/420integrated/fourtwentyjs-devp2p)

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

This library bundles different components for lower-level peer-to-peer connection and message exchange:

- Distributed Peer Table (DPT) / Node Discovery
- RLPx Transport Protocol
- 420coin Wire Protocol (FOURTWENTY)
- Light 420coin Subprotocol (LES/2)


## Run/Build

This library has to be compiled with typescript to a `Node 6` and above friendly source format.
For triggering a (first) build to create the `dist/` directory run:

```
npm run build
```

You can also use `ts-node` to run a script without first transpiling to js (you need to `npm i --save-dev ts-node` first):

```
node -r ts-node/register [YOUR_SCRIPT_TO_RUN.ts]
```

## Usage/Examples

All components of this library are implemented as Node `EventEmitter` objects
and make heavy use of the Node.js network stack.

You can react on events from the network like this:

```
dpt.on('peer:added', (peer) => {
  // Do something...
})
```

Basic example to connect to some bootstrap nodes and get basic peer info:

- [simple](examples/simple.ts)

Communicate with peers to read new transaction and block information:

- [peer-communication](examples/peer-communication.ts)

Run an example with:

```
DEBUG=devp2p:* node -r ts-node/register ./examples/peer-communication.ts
```

## Distributed Peer Table (DPT) / Node Discovery

Maintain/manage a list of peers, see [./src/dpt/](./src/dpt/), also
includes node discovery ([./src/dpt/server.ts](./src/dpt/server.ts))

### Usage

Create your peer table:

```
const dpt = new DPT(Buffer.from(PRIVATE_KEY, 'hex'), {
  endpoint: {
    address: '0.0.0.0',
    udpPort: null,
    tcpPort: null
  }
})
```

Add some bootstrap nodes (or some custom nodes with `dpt.addPeer()`):

```
dpt.bootstrap(bootnode).catch((err) => console.error('Something went wrong!'))
```

### API

See the following [diagram](./diagram/devp2p.svg) for a high level overview on the library.

#### `DPT` (extends `EventEmitter`)

Distributed Peer Table. Manages a Kademlia DHT K-bucket (`Kbucket`) for storing peer information
and a `BanList` for keeping a list of bad peers. `Server` implements the node discovery (`ping`,
`pong`, `findNeighbours`).

##### `new DPT(privateKey, options)`

Creates new DPT object

- `privateKey` - Key for message encoding/signing.
- `options.refreshInterval` - Interval in ms for refreshing (calling `findNeighbours`) the peer list (default: `60s`).
- `options.createSocket` - A datagram (dgram) `createSocket` function, passed to `Server` (default: `dgram.createSocket.bind(null, 'udp4')`).
- `options.timeout` - Timeout in ms for server `ping`, passed to `Server` (default: `10s`).
- `options.endpoint` - Endpoint information to send with the server `ping`, passed to `Server` (default: `{ address: '0.0.0.0', udpPort: null, tcpPort: null }`).

#### `dpt.bootstrap(peer)` (`async`)

Uses a peer as new bootstrap peer and calls `findNeighbouts`.

- `peer` - Peer to be added, format `{ address: [ADDRESS], udpPort: [UDPPORT], tcpPort: [TCPPORT] }`.

#### `dpt.addPeer(object)` (`async`)

Adds a new peer.

- `object` - Peer to be added, format `{ address: [ADDRESS], udpPort: [UDPPORT], tcpPort: [TCPPORT] }`.

For other utility functions like `getPeer`, `getPeers` see [./src/dpt/dpt.ts](./src/dpt/dpt.ts).

### Events

Events emitted:

| Event        |         Description          |
| ------------ | :--------------------------: |
| peer:added   |   Peer added to DHT bucket   |
| peer:removed | Peer removed from DHT bucket |
| peer:new     |        New peer added        |
| listening    |    Forwarded from server     |
| close        |    Forwarded from server     |
| error        |    Forwarded from server     |

### Reference

- [Node discovery protocol](https://github.com/420integrated/go-420coin/wiki/wiki/Node-discovery-protocol)
- [RLPx - Node Discovery Protocol](https://github.com/420integrated/go-420coin/devp2p/blob/master/rlpx.md#node-discovery)
- [Kademlia Peer Selection](https://github.com/420integrated/go-420coin/wiki/wiki/Kademlia-Peer-Selection)

## RLPx Transport Protocol

Connect to a peer, organize the communication, see [./src/rlpx/](./src/rlpx/)

### Usage

Create your `RLPx` object, e.g.:

```
const rlpx = new devp2p.RLPx(PRIVATE_KEY, {
  dpt: dpt,
  maxPeers: 25,
  capabilities: [
    devp2p.FOURTWENTY.fourtwenty63,
    devp2p.FOURTWENTY.fourtwenty62
  ],
  listenPort: null
})
```

### API

#### `RLPx` (extends `EventEmitter`)

Manages the handshake (`ECIES`) and the handling of the peer communication (`Peer`).

##### `new RLPx(privateKey, options)`

Creates new RLPx object

- `privateKey` - Key for message encoding/signing.
- `options.timeout` - Peer `ping` timeout in ms (default: `10s`).
- `options.maxPeers` - Max number of peer connections (default: `10`).
- `options.clientId` - Client ID string (default example: `fourtwentyjs-devp2p/v2.1.3/darwin-x64/nodejs`).
- `options.remoteClientIdFilter` - Optional list of client ID filter strings (e.g. `['go1.5', 'quorum']`).
- `options.capabilities` - Upper layer protocol capabilities, e.g. `[devp2p.FOURTWENTY.fourtwenty63, devp2p.FOURTWENTY.fourtwenty62]`.
- `options.listenPort` - The listening port for the server or `null` for default.
- `options.dpt` - `DPT` object for the peers to connect to (default: `null`, no `DPT` peer management).

#### `rlpx.connect(peer)` (`async`)

Manually connect to peer without `DPT`.

- `peer` - Peer to connect to, format `{ id: PEER_ID, address: PEER_ADDRESS, port: PEER_PORT }`.

For other connection/utility functions like `listen`, `getPeers` see [./src/rlpx/rlpx.ts](./src/rlpx/rlpx.ts).

### Events

Events emitted:

| Event        |          Description           |
| ------------ | :----------------------------: |
| peer:added   | Handshake with peer successful |
| peer:removed |     Disconnected from peer     |
| peer:error   |    Error connecting to peer    |
| listening    |     Forwarded from server      |
| close        |     Forwarded from server      |
| error        |     Forwarded from server      |

### Reference

- [RLPx: Cryptographic Network & Transport Protocol](https://github.com/420integrated/go-420coin/devp2p/blob/master/rlpx.md)
- [devp2p wire protocol](https://github.com/420integrated/go-420coin/wiki/wiki/%C3%90%CE%9EVp2p-Wire-Protocol)

## 420coin Wire Protocol (FOURTWENTY)

Upper layer protocol for exchanging 420coin network data like block headers or transactions with a node, see [./src/fourtwenty/](./src/fourtwenty/).

### Usage

Send the initial status message with `sendStatus()`, then wait for the corresponding `status` message
to arrive to start the communication.

```
fourtwenty.once('status', () => {
  // Send an initial message
  fourtwenty.sendMessage()
})
```

Wait for follow-up messages to arrive, send your responses.

```
fourtwenty.on('message', async (code, payload) => {
  if (code === devp2p.FOURTWENTY.MESSAGE_CODES.NEW_BLOCK_HASHES) {
    // Do something with your new block hashes :-)
  }
})
```

See the `peer-communication.ts` example for a more detailed use case.

### API

#### `FOURTWENTY` (extends `EventEmitter`)

Handles the different message types like `NEW_BLOCK_HASHES` or `GET_NODE_DATA` (see `MESSAGE_CODES`) for
a complete list. Currently protocol versions `PV62` and `PV63` are supported.

##### `new FOURTWENTY(privateKey, options)`

Normally not instantiated directly but created as a `SubProtocol` in the `Peer` object.

- `version` - The protocol version for communicating, e.g. `63`.
- `peer` - `Peer` object to communicate with.
- `send` - Wrapped `peer.sendMessage()` function where the communication is routed to.

#### `fourtwenty.sendStatus(status)`

Send initial status message.

- `status` - Status message to send, format `{ networkId: CHAIN_ID, td: TOTAL_DIFFICULTY_BUFFER, bestHash: BEST_HASH_BUFFER, genesisHash: GENESIS_HASH_BUFFER }`.

#### `fourtwenty.sendMessage(code, payload)`

Send initial status message.

- `code` - The message code, see `MESSAGE_CODES` for available message types.
- `payload` - Payload as a list, will be rlp-encoded.

### Events

Events emitted:

| Event   |     Description      |
| ------- | :------------------: |
| message |   Message received   |
| status  | Status info received |

### Reference

- [420coin wire protocol](https://github.com/420integrated/go-420coin/wiki/wiki/420coin-Wire-Protocol)

## Light 420coin Subprotocol (LES)

Upper layer protocol used by light clients, see [./src/les/](./src/les/).

### Usage

Send the initial status message with `sendStatus()`, then wait for the corresponding `status` message
to arrive to start the communication.

```
les.once('status', () => {
  // Send an initial message
  les.sendMessage()
})
```

Wait for follow-up messages to arrive, send your responses.

```
les.on('message', async (code, payload) => {
  if (code === devp2p.LES.MESSAGE_CODES.BLOCK_HEADERS) {
    // Do something with your new block headers :-)
  }
})
```

See the `peer-communication-les.ts` example for a more detailed use case.

### API

#### `LES` (extends `EventEmitter`)

Handles the different message types like `BLOCK_HEADERS` or `GET_PROOFS_V2` (see `MESSAGE_CODES`) for
a complete list. Currently protocol version `LES/2` running in client-mode is supported.

##### `new LES(privateKey, options)`

Normally not instantiated directly but created as a `SubProtocol` in the `Peer` object.

- `version` - The protocol version for communicating, e.g. `2`.
- `peer` - `Peer` object to communicate with.
- `send` - Wrapped `peer.sendMessage()` function where the communication is routed to.

#### `les.sendStatus(status)`

Send initial status message.

- `status` - Status message to send, format `{ networkId: CHAIN_ID, headTd: TOTAL_DIFFICULTY_BUFFER, headHash: HEAD_HASH_BUFFER, headNum: HEAD_NUM_BUFFER, genesisHash: GENESIS_HASH_BUFFER }`.

#### `les.sendMessage(code, reqId, payload)`

Send initial status message.

- `code` - The message code, see `MESSAGE_CODES` for available message types.
- `reqId` - Request ID, will be echoed back on response.
- `payload` - Payload as a list, will be rlp-encoded.

### Events

Events emitted:

| Event   |     Description      |
| ------- | :------------------: |
| message |   Message received   |
| status  | Status info received |

### Reference

- [Light client protocol](https://github.com/420integrated/go-420coin/wiki/wiki/Light-client-protocol)

## Tests

There are unit tests in the `test/` directory which can be run with:

```
npm run test
```

## Debugging

This library uses [debug](https://github.com/visionmedia/debug) debugging utility package.

For the debugging output to show up, set the `DEBUG` environment variable (e.g. in Linux/Mac OS:
`export DEBUG=*,-babel`).

Use the `DEBUG` environment variable to active the logger output you are interested in, e.g.:

DEBUG=devp2p:dpt:\*,devp2p:fourtwenty node -r ts-node/register [YOUR_SCRIPT_TO_RUN.ts]

For more verbose output on logging (e.g. to output the entire msg payload) use the `verbose` logger
in addition:

DEBUG=devp2p:dpt:\*,devp2p:fourtwenty,verbose node -r ts-node/register [YOUR_SCRIPT_TO_RUN.ts]

Exemplary logging output:

```
Add peer: 52.3.158.184:13013 G420/v1.7.3-unstable-479aa61f/linux-amd64/go1.9 (fourtwenty63) (total: 2)
  devp2p:rlpx:peer Received body 52.169.42.101:13013 01c110 +133ms
  devp2p:rlpx:peer Message code: 1 - 0 = 1 +0ms
  devp2p:rlpx refill connections.. queue size: 0, open slots: 20 +1ms
  devp2p:rlpx 52.169.42.101:13013 disconnect, reason: 16 +1ms
Remove peer: 52.169.42.101:13013 (peer disconnect, reason code: 16) (total: 1)
```

## Developer

### Diagram Updates

To update the structure diagram files in the root folder open the `devp2p.drawio` file in [draw.io](https://draw.io/), make your changes, and open a PR with the updated files. Export `svg` and `png` with `border` `width=20` and `transparency=false`. For `png` go to "Advanced" and select `300 DPI`.

## General References

### Other Implementations

The following is a list of major implementations of the `devp2p` stack in other languages:

- Python: [pydevp2p](https://github.com/420integrated/go-420coin/pydevp2p)
- Go: [Go 420coin](https://github.com/420integrated/go-420coin/tree/master/p2p)
- Elixir: [Exthereum](https://github.com/exthereum/exth_crypto)

### Links

- [Blog article series](https://ocalog.com/post/10/) on implementing 420coin protocol stack

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.0.3] - 2020-09-29

- Moved `TypeScript` type packages for `lru-cache` and `bl` from `devDependencies` to
  `dependencies`, PR [#90](https://github.com/420integrated/fourtwentyjs-devp2p/pull/90)

[3.0.3]: https://github.com/420integrated/fourtwentyjs-devp2p/compare/v3.0.2...v3.0.3

## [3.0.2] - 2020-09-28

- Fixed `TypeScript` import issue causing problems when integrating the library in a
  `TypeScript` project, PR [#88](https://github.com/420integrated/fourtwentyjs-devp2p/pull/88)
- Updated `k-bucket` library to `v5`, added types from new `@types/k-bucket` package from
  @tomonari-t, PR [#88](https://github.com/420integrated/fourtwentyjs-devp2p/pull/88)

[3.0.2]: https://github.com/420integrated/fourtwentyjs-devp2p/compare/v3.0.1...v3.0.2

## [3.0.1] - 2020-06-10

This release focuses on improving the [debugging](https://github.com/420integrated/fourtwentyjs-devp2p#debugging)
capabilities of the library. PR [#72](https://github.com/420integrated/fourtwentyjs-devp2p/pull/72)
reduces the **verbosity** of the log output to cut on noise on everyday debugging. There is a new `verbose`
logger to retain the more verbose output (e.g. with full message bodies) which can be used like this:

```shell
DEBUG=devp2p:*,verbose node -r ts-node/register ./examples/peer-communication.ts
```

**Other Logging Improvements**

Relevant PRs [#75](https://github.com/420integrated/fourtwentyjs-devp2p/pull/75) and
[#73](https://github.com/420integrated/fourtwentyjs-devp2p/pull/73):

- Added number of peers to `refillConnections()` debug message
- Replaced try/catch logic for EIP-8 auth check to avoid side-effects and get rid of misleading _wrong-ecies-header_ debug output
- Moved debug output in `BanList.add()` after the set operation to get the correct size output
- Added debug message for `DISCONNECT` reason from peer (this was always some constant re-debug reason, and at the end it's mostly `TOO_MANY_PEERS`)
- Internalize detached logger output from the `devp2p:util` logger

**Other Changes**

- Refactored `Peer` class for better code readability, PR [#77](https://github.com/420integrated/fourtwentyjs-devp2p/pull/77)

There has also been a new [high-level diagram](https://github.com/420integrated/fourtwentyjs-devp2p#api) added to the `README` which can be used to get an overview on the structure, available loggers and the event flow of the library (PR [#76](https://github.com/420integrated/fourtwentyjs-devp2p/pull/76)).

[3.0.1]: https://github.com/420integrated/fourtwentyjs-devp2p/compare/v3.0.0...v3.0.1

## [3.0.0] - 2020-05-25

First `TypeScript` release of the library, see PR [#56](https://github.com/420integrated/fourtwentyjs-devp2p/pull/56) for all the changes and associated discussion.

All source parts of the library have been ported to `TypeScript` and working with the library should now therefore be much more reliable due to the additional type safety features provided by the `TypeScript` language. The API of the library remains unchanged in a `JavaScript` context.

**Noteworthy Changes from PR [#56](https://github.com/420integrated/fourtwentyjs-devp2p/pull/56):**

- Type additions for all method signatures and class members of all protocol components (`dpt`, `fourtwenty`, `les`, `rlpx`)
- Addition of various structuring interfaces (like [PeerInfo](https://github.com/420integrated/fourtwentyjs-devp2p/blob/master/src/dpt/message.ts#L10) for `DPT` message input) and `enum` constructs (like [MESSAGE_CODES](https://github.com/420integrated/fourtwentyjs-devp2p/blob/master/src/eth/index.ts#L186) from the `FOURTWENTY` protocol)
- Port of the [examples](https://github.com/420integrated/fourtwentyjs-devp2p/tree/master/examples) to `TypeScript`
- Port of all the [test cases](https://github.com/420integrated/fourtwentyjs-devp2p/tree/master/test) to `TypeScript`
- Integration of the library into the common [fourtwentyjs-config](https://github.com/420integrated/fourtwentyjs-config) 420coinJS configuration setup (`standard` -> `TSLint` linting, docs with `TypeDoc`, `TypeScript` compilation, `Prettier` formatting rules)
- Lots of code cleanups and code part modernizations

Thanks @dryajov for all the great work on this! ❤

**Other Updates:**

- Added Node 12,13 support, upgrade from Travis to GitHub actions, PR [#57](https://github.com/420integrated/fourtwentyjs-devp2p/pull/57)
- Updated `fourtwentyjs-common` dependency to `v1.5.1` for a bootnode update, PR [#67](https://github.com/420integrated/fourtwentyjs-devp2p/pull/67)
- Removed Node 6, 8 support, updated `secp256k1` dependency to from `v3.1.0` to `v4.0.1`, PR [#68](https://github.com/420integrated/fourtwentyjs-devp2p/pull/68)
- Updated `keccak` dependency to `v3.0.0`, PR [#64](https://github.com/420integrated/fourtwentyjs-devp2p/pull/64)
- Some dependency cleanup, PRs [#62](https://github.com/420integrated/fourtwentyjs-devp2p/pull/62), [#65](https://github.com/420integrated/fourtwentyjs-devp2p/pull/65), [#58](https://github.com/420integrated/fourtwentyjs-devp2p/pull/58)

[3.0.0]: https://github.com/420integrated/fourtwentyjs-devp2p/compare/v2.5.1...v3.0.0

## [2.5.1] - 2018-12-12

- Fix connection error by ignoring `RLPX` peers with missing tcp port, PR [#45](https://github.com/420integrated/fourtwentyjs-devp2p/pull/45)

[2.5.1]: https://github.com/420integrated/fourtwentyjs-devp2p/compare/v2.5.0...v2.5.1

## [2.5.0] - 2018-03-22

- Light client protocol (`LES/2`) implementation, PR [#21](https://github.com/420integrated/fourtwentyjs-devp2p/pull/21)
- `LES/2` usage example, see: `examples/peer-communication-les.js`
- Better test coverage for upper-layer protocols (`FOURTWENTY`, `LES/2`), PR [#34](https://github.com/420integrated/fourtwentyjs-devp2p/pull/34)


examples/peer-communication.ts
import * as devp2p from '../src/index'
import { FOURTWENTY, Peer } from '../src/index'
import Tx from 'fourtwentyjs-tx'
import Block from 'fourtwentyjs-block'
import LRUCache from 'lru-cache'
import ms from 'ms'
import chalk from 'chalk'
import assert from 'assert'
import { randomBytes } from 'crypto'
import rlp from 'rlp-encoding'

const PRIVATE_KEY = randomBytes(32)
const CHAIN_ID = 1

const Common = require('fourtwentyjs-common').default
const config = new Common('mainnet')
const bootstrapNodes = config.bootstrapNodes()
const BOOTNODES = bootstrapNodes.map((node: any) => {
  return {
    address: node.ip,
    udpPort: node.port,
    tcpPort: node.port,
  }
})
const REMOTE_CLIENTID_FILTER = [
  'go1.5',
  'go1.6',
  'go1.7',
  'quorum',
  'pirl',
  'ubiq',
  'gmc',
  'gwhale',
  'prichain',
]

const CHECK_BLOCK_TITLE = 'Byzantium Fork' // Only for debugging/console output
const CHECK_BLOCK_NR = 4370000
const CHECK_BLOCK = 'b1fcff633029ee18ab6482b58ff8b6e95dd7c82a954c852157152a7a6d32785e'
const CHECK_BLOCK_HEADER = rlp.decode(
  Buffer.from(
    'f9020aa0a0890da724dd95c90a72614c3a906e402134d3859865f715f5dfb398ac00f955a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347942a65aca4d5fc5b5c859090a6c34d164135398226a074cccff74c5490fbffc0e6883ea15c0e1139e2652e671f31f25f2a36970d2f87a00e750bf284c2b3ed1785b178b6f49ff3690a3a91779d400de3b9a3333f699a80a0c68e3e82035e027ade5d966c36a1d49abaeec04b83d64976621c355e58724b8bb90100040019000040000000010000000000021000004020100688001a05000020816800000010a0000100201400000000080100020000000400080000800004c0200000201040000000018110400c000000200001000000280000000100000010010080000120010000050041004000018000204002200804000081000011800022002020020140000000020005080001800000000008102008140008600000000100000500000010080082002000102080000002040120008820400020100004a40801000002a0040c000010000114000000800000050008300020100000000008010000000100120000000040000000808448200000080a00000624013000000080870552416761fabf83475b02836652b383661a72845a25c530894477617266506f6f6ca0dc425fdb323c469c91efac1d2672dfdd3ebfde8fa25d68c1b3261582503c433788c35ca7100349f430',
    'hex',
  ),
)

const getPeerAddr = (peer: Peer) => `${peer._socket.remoteAddress}:${peer._socket.remotePort}`

// DPT
const dpt = new devp2p.DPT(PRIVATE_KEY, {
  refreshInterval: 30000,
  endpoint: {
    address: '0.0.0.0',
    udpPort: null,
    tcpPort: null,
  },
})

dpt.on('error', err => console.error(chalk.red(`DPT error: ${err}`)))

// RLPx
const rlpx = new devp2p.RLPx(PRIVATE_KEY, {
  dpt: dpt,
  maxPeers: 25,
  capabilities: [devp2p.FOURTWENTY.fourtwenty63, devp2p.FOURTWENTY.fourtwenty62],
  remoteClientIdFilter: REMOTE_CLIENTID_FILTER,
  listenPort: null,
})

rlpx.on('error', err => console.error(chalk.red(`RLPx error: ${err.stack || err}`)))

rlpx.on('peer:added', peer => {
  const addr = getPeerAddr(peer)
  const fourtwenty = peer.getProtocols()[0]
  const requests: {
    headers: any[]
    bodies: any[]
    msgTypes: { [key: string]: FOURTWENTY.MESSAGE_CODES }
  } = { headers: [], bodies: [], msgTypes: {} }

  const clientId = peer.getHelloMessage().clientId
  console.log(
    chalk.green(
      `Add peer: ${addr} ${clientId} (fourtwenty${fourtwenty.getVersion()}) (total: ${rlpx.getPeers().length})`,
    ),
  )

  fourtwenty.sendStatus({
    networkId: CHAIN_ID,
    td: devp2p.int2buffer(17179869184), // total difficulty in genesis block
    bestHash: Buffer.from(
      'd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3',
      'hex',
    ),
    genesisHash: Buffer.from(
      'd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3',
      'hex',
    ),
  })

  // check CHECK_BLOCK
  let forkDrop: NodeJS.Timeout
  let forkVerified = false
  fourtwenty.once('status', () => {
    fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.GET_BLOCK_HEADERS, [CHECK_BLOCK_NR, 1, 0, 0])
    forkDrop = setTimeout(() => {
      peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
    }, ms('15s'))
    peer.once('close', () => clearTimeout(forkDrop))
  })

  fourtwenty.on('message', async (code: FOURTWENTY.MESSAGE_CODES, payload: any) => {
    if (code in FOURTWENTY.MESSAGE_CODES) {
      requests.msgTypes[code] = code + 1
    } else {
      requests.msgTypes[code] = 1
    }

    switch (code) {
      case devp2p.FOURTWENTY.MESSAGE_CODES.NEW_BLOCK_HASHES:
        if (!forkVerified) break

        for (let item of payload) {
          const blockHash = item[0]
          if (blocksCache.has(blockHash.toString('hex'))) continue
          setTimeout(() => {
            fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.GET_BLOCK_HEADERS, [blockHash, 1, 0, 0])
            requests.headers.push(blockHash)
          }, ms('0.1s'))
        }
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.TX:
        if (!forkVerified) break

        for (let item of payload) {
          const tx = new Tx(item)
          if (isValidTx(tx)) onNewTx(tx, peer)
        }

        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.GET_BLOCK_HEADERS:
        const headers = []
        // hack
        if (devp2p.buffer2int(payload[0]) === CHECK_BLOCK_NR) {
          headers.push(CHECK_BLOCK_HEADER)
        }

        if (requests.headers.length === 0 && requests.msgTypes[code] >= 8) {
          peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
        } else {
          fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.BLOCK_HEADERS, headers)
        }
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.BLOCK_HEADERS:
        if (!forkVerified) {
          if (payload.length !== 1) {
            console.log(
              `${addr} expected one header for ${CHECK_BLOCK_TITLE} verify (received: ${payload.length})`,
            )
            peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
            break
          }

          const expectedHash = CHECK_BLOCK
          const header = new Block.Header(payload[0])
          if (header.hash().toString('hex') === expectedHash) {
            console.log(`${addr} verified to be on the same side of the ${CHECK_BLOCK_TITLE}`)
            clearTimeout(forkDrop)
            forkVerified = true
          }
        } else {
          if (payload.length > 1) {
            console.log(
              `${addr} not more than one block header expected (received: ${payload.length})`,
            )
            break
          }

          let isValidPayload = false
          const header = new Block.Header(payload[0])
          while (requests.headers.length > 0) {
            const blockHash = requests.headers.shift()
            if (header.hash().equals(blockHash)) {
              isValidPayload = true
              setTimeout(() => {
                fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.GET_BLOCK_BODIES, [blockHash])
                requests.bodies.push(header)
              }, ms('0.1s'))
              break
            }
          }

          if (!isValidPayload) {
            console.log(`${addr} received wrong block header ${header.hash().toString('hex')}`)
          }
        }

        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.GET_BLOCK_BODIES:
        if (requests.headers.length === 0 && requests.msgTypes[code] >= 8) {
          peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
        } else {
          fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.BLOCK_BODIES, [])
        }
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.BLOCK_BODIES:
        if (!forkVerified) break

        if (payload.length !== 1) {
          console.log(`${addr} not more than one block body expected (received: ${payload.length})`)
          break
        }

        let isValidPayload = false
        while (requests.bodies.length > 0) {
          const header = requests.bodies.shift()
          const block = new Block([header.raw, payload[0][0], payload[0][1]])
          const isValid = await isValidBlock(block)
          if (isValid) {
            isValidPayload = true
            onNewBlock(block, peer)
            break
          }
        }

        if (!isValidPayload) {
          console.log(`${addr} received wrong block body`)
        }

        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.NEW_BLOCK:
        if (!forkVerified) break

        const newBlock = new Block(payload[0])
        const isValidNewBlock = await isValidBlock(newBlock)
        if (isValidNewBlock) onNewBlock(newBlock, peer)

        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.GET_NODE_DATA:
        if (requests.headers.length === 0 && requests.msgTypes[code] >= 8) {
          peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
        } else {
          fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.NODE_DATA, [])
        }
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.NODE_DATA:
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.GET_RECEIPTS:
        if (requests.headers.length === 0 && requests.msgTypes[code] >= 8) {
          peer.disconnect(devp2p.DISCONNECT_REASONS.USELESS_PEER)
        } else {
          fourtwenty.sendMessage(devp2p.FOURTWENTY.MESSAGE_CODES.RECEIPTS, [])
        }
        break

      case devp2p.FOURTWENTY.MESSAGE_CODES.RECEIPTS:
        break
    }
  })
})

rlpx.on('peer:removed', (peer, reasonCode, disconnectWe) => {
  const who = disconnectWe ? 'we disconnect' : 'peer disconnect'
  const total = rlpx.getPeers().length
  console.log(
    chalk.yellow(
      `Remove peer: ${getPeerAddr(peer)} - ${who}, reason: ${peer.getDisconnectPrefix(
        reasonCode,
      )} (${String(reasonCode)}) (total: ${total})`,
    ),
  )
})

rlpx.on('peer:error', (peer, err) => {
  if (err.code === 'ECONNRESET') return

  if (err instanceof assert.AssertionError) {
    const peerId = peer.getId()
    if (peerId !== null) dpt.banPeer(peerId, ms('5m'))

    console.error(chalk.red(`Peer error (${getPeerAddr(peer)}): ${err.message}`))
    return
  }

  console.error(chalk.red(`Peer error (${getPeerAddr(peer)}): ${err.stack || err}`))
})

// uncomment, if you want accept incoming connections
// rlpx.listen(13013, '0.0.0.0')
// dpt.bind(13013, '0.0.0.0')

for (let bootnode of BOOTNODES) {
  dpt.bootstrap(bootnode).catch(err => {
    console.error(chalk.bold.red(`DPT bootstrap error: ${err.stack || err}`))
  })
}

// connect to local 420coin node (debug)
/*
dpt.addPeer({ address: '127.0.0.1', udpPort: 13013, tcpPort: 13013 })
  .then((peer) => {
    return rlpx.connect({
      id: peer.id,
      address: peer.address,
      port: peer.tcpPort
    })
  })
  .catch((err) => console.log(`error on connection to local node: ${err.stack || err}`))
*/

const txCache = new LRUCache({ max: 1000 })
function onNewTx(tx: Tx, peer: Peer) {
  const txHashHex = tx.hash().toString('hex')
  if (txCache.has(txHashHex)) return

  txCache.set(txHashHex, true)
  console.log(`New tx: ${txHashHex} (from ${getPeerAddr(peer)})`)
}

const blocksCache = new LRUCache({ max: 100 })
function onNewBlock(block: Block, peer: Peer) {
  const blockHashHex = block.hash().toString('hex')
  const blockNumber = devp2p.buffer2int(block.header.number)
  if (blocksCache.has(blockHashHex)) return

  blocksCache.set(blockHashHex, true)
  console.log(
    `----------------------------------------------------------------------------------------------------------`,
  )
  console.log(`New block ${blockNumber}: ${blockHashHex} (from ${getPeerAddr(peer)})`)
  console.log(
    `----------------------------------------------------------------------------------------------------------`,
  )
  for (let tx of block.transactions) onNewTx(tx, peer)
}

function isValidTx(tx: Tx) {
  return tx.validate(false)
}

async function isValidBlock(block: Block) {
  if (!block.validateUnclesHash()) return false
  if (!block.transactions.every(isValidTx)) return false
  return new Promise((resolve, reject) => {
    block.genTxTrie(() => {
      try {
        resolve(block.validateTransactionsTrie())
      } catch (err) {
        reject(err)
      }
    })
  })
}

setInterval(() => {
  const peersCount = dpt.getPeers().length
  const openSlots = rlpx._getOpenSlots()
  const queueLength = rlpx._peersQueue.length
  const queueLength2 = rlpx._peersQueue.filter(o => o.ts <= Date.now()).length

  console.log(
    chalk.yellow(
      `Total nodes in DPT: ${peersCount}, open slots: ${openSlots}, queue: ${queueLength} / ${queueLength2}`,
    ),
  )
}, ms('30s'))
